### DISCUSSION TITLE
Next.js Dependency Warnings with Migrator

### DISCUSSION THREAD

**Discussion by maphe** [mikro-orm/mikro-orm#5486] [Q&A]
Hi, I'm playing around and trying to integrate Mikro ORM in a new Next.js project (App Router) and I'm running into some warnings. If I understand them correctly they seem to come from Mikro ORM dependencies, rather than Mikro ORM itself, in combination with the next.js compiler. But still I thought I'd ask here if anybody else ran into it and/or if there's a fix that I haven't found.

Here's the console output when the Page compiles in dev mode:
```
 ⚠ ../../node_modules/jju/lib/utils.js
Critical dependency: require function is used in a way in which dependencies cannot be statically extracted

Import trace for requested module:
../../node_modules/jju/lib/utils.js
../../node_modules/jju/index.js
../../node_modules/@rushstack/node-core-library/lib/JsonFile.js
../../node_modules/@rushstack/node-core-library/lib/index.js
../../node_modules/@rushstack/terminal/lib/NormalizeNewlinesTextRewriter.js
../../node_modules/@rushstack/terminal/lib/index.js
../../node_modules/@rushstack/ts-command-line/lib/providers/CommandLineParser.js
../../node_modules/@rushstack/ts-command-line/lib/index.js
../../node_modules/umzug/lib/cli.js
../../node_modules/umzug/lib/index.js
../../node_modules/@mikro-orm/migrations/Migrator.js
../../node_modules/@mikro-orm/migrations/index.js
../../node_modules/@mikro-orm/migrations/index.mjs
../../libs/db/src/mikro-orm.config.ts
../../libs/db/src/db.ts
../../libs/db/src/index.ts
./src/app/page.tsx

../../node_modules/umzug/lib/umzug.js
Critical dependency: the request of a dependency is an expression

Import trace for requested module:
../../node_modules/umzug/lib/umzug.js
../../node_modules/umzug/lib/index.js
../../node_modules/@mikro-orm/migrations/Migrator.js
../../node_modules/@mikro-orm/migrations/index.js
../../node_modules/@mikro-orm/migrations/index.mjs
../../libs/db/src/mikro-orm.config.ts
../../libs/db/src/db.ts
../../libs/db/src/index.ts
./src/app/page.tsx
[info] MikroORM version: 6.2.2
```

I'm running:
- MikroORM `6.2.2`
- Next.js `14.2.2`
- Nx `18.3.3` (integrated monorepo)

As this looks to be triggered by Migrator, maybe there's a way to disable the Migrator when running the app, and having it on only when running migration commands?

**Comment by B4nan** [comment]
Looking at the stack trace, the import is coming from the ORM config, no? 

Some notes about how to deal with bundlers are here https://mikro-orm.io/docs/deployment

And indeed, things like this can be usually worked around by changing the bundler config - don't expect those things to be fixable from the other end, this needs to be fixed in your next project (probably in its webpack config if that's what they use). I am no webpack expert, can't help much with those things.

**Comment by maphe** [comment]
Yeah it's coming from the config, the Migrator extension. When I remove it the warning goes away. I'll try to do some webpack tuning and see if I can manage something. Meanwhile I'm curious what you think about the following workaround.

A lighter config for the app to load (`MikroOrm.init(configMinimal)`), that doesn't care about the extensions:
```ts
const configMinimal: Options = {
  driver: PostgreSqlDriver,
  dbName: 'demo_db',
  clientUrl: 'postgresql://localhost:5432',
  entities: [/*...*/],
  metadataProvider: ReflectMetadataProvider,
  discovery: { disableDynamicFileAccess: true },
  debug: true,
};
```
And the complete config that gets loaded by the CLI when running migrations
```ts
const config: Options = {
  ...configMinimal,
  migrations: {
    path: process.cwd() + '/migrations',
    transactional: true,
    disableForeignKeys: true,
    allOrNothing: true,
    emit: 'ts',
  },
  extensions: [SeedManager, Migrator],
};
```

**Reply by B4nan** [reply]
It still sounds weird that this would trigger a build error. The whole point of those ORM extensions is to let you import stuff instead of using dynamic imports from the ORM's code, which are problematic for this exact reason. Maybe you have the migrations package declared only as a dev dependency? Just a guess, not sure if this can be problematic or not.

I think it's fine to use a config extension just for the CLI, you can even keep the migrations config in there. I'd guess the problem won't be the values in your config, but the import (of the migrations package) inside it.

**Reply by maphe** [reply]
Migration package is well imported in the non-dev dependencies so it doesn't look to be the problem.

The Webpack section of the documentation has too many things that I'm not sure to understand tbh, I tried a few things but it only made it worse.

I'm splitting the config into 2 files as a simple workaround for now. Maybe I'll revisit later.
